File: implementations/typescript/src/eval/pawnStructure.ts
import { Board } from "../board";
import { Color, PieceType, Square } from "../types";

const PASSED_PAWN_BONUS: number[] = [0, 10, 20, 40, 60, 90, 120, 0];
const DOUBLED_PAWN_PENALTY = -20;
const ISOLATED_PAWN_PENALTY = -15;
const BACKWARD_PAWN_PENALTY = -10;
const CONNECTED_PAWN_BONUS = 5;
const PAWN_CHAIN_BONUS = 10;

export function evaluate(board: Board): number {
  let score = 0;
  
  score += evaluateColor(board, "white");
  score -= evaluateColor(board, "black");
  
  return score;
}

function evaluateColor(board: Board, color: Color): number {
  let score = 0;
  const pawnFiles: number[] = new Array(8).fill(0);
  const pawnPositions: [Square, number, number][] = [];
  
  for (let square = 0; square < 64; square++) {
    const piece = board.getPiece(square);
    if (piece && piece.color === color && piece.type === "P") {
      const file = square % 8;
      const rank = Math.floor(square / 8);
      pawnFiles[file]++;
      pawnPositions.push([square, rank, file]);
    }
  }
  
  for (const [square, rank, file] of pawnPositions) {
    if (pawnFiles[file] > 1) {
      score += DOUBLED_PAWN_PENALTY;
    }
    
    if (isIsolated(file, pawnFiles)) {
      score += ISOLATED_PAWN_PENALTY;
    }
    
    if (isPassed(board, square, rank, file, color)) {
      const bonusRank = color === "white" ? rank : 7 - rank;
      score += PASSED_PAWN_BONUS[bonusRank];
    }
    
    if (isConnected(board, square, file, color)) {
      score += CONNECTED_PAWN_BONUS;
    }
    
    if (isInChain(board, square, rank, file, color)) {
      score += PAWN_CHAIN_BONUS;
    }
    
    if (isBackward(board, square, rank, file, color, pawnFiles)) {
      score += BACKWARD_PAWN_PENALTY;
    }
  }
  
  return score;
}

function isIsolated(file: number, pawnFiles: number[]): boolean {
  const leftFile = file > 0 ? pawnFiles[file - 1] : 0;
  const rightFile = file < 7 ? pawnFiles[file + 1] : 0;
  return leftFile === 0 && rightFile === 0;
}

function isPassed(board: Board, square: Square, rank: number, file: number, color: Color): boolean {
  const startRank = color === "white" ? rank + 1 : 0;
  const endRank = color === "white" ? 8 : rank;
  const direction = color === "white" ? 1 : -1;
  
  for (let checkFile = Math.max(0, file - 1); checkFile <= Math.min(7, file + 1); checkFile++) {
    let currentRank = startRank;
    
    while (true) {
      if (color === "white") {
        if (currentRank >= endRank) break;
      } else {
        if (currentRank >= rank) break;
      }
      
      const checkSquare = currentRank * 8 + checkFile;
      const piece = board.getPiece(checkSquare);
      if (piece && piece.type === "P" && piece.color !== color) {
        return false;
      }
      
      currentRank = direction > 0 ? currentRank + 1 : Math.max(0, currentRank - 1);
      if (direction < 0 && currentRank === 0) break;
    }
  }
  
  return true;
}

function isConnected(board: Board, square: Square, file: number, color: Color): boolean {
  const rank = Math.floor(square / 8);
  
  const adjacentFiles = [Math.max(0, file - 1), Math.min(7, file + 1)];
  for (const adjacentFile of adjacentFiles) {
    if (adjacentFile !== file) {
      const adjacentSquare = rank * 8 + adjacentFile;
      const piece = board.getPiece(adjacentSquare);
      if (piece && piece.color === color && piece.type === "P") {
        return true;
      }
    }
  }
  
  return false;
}

function isInChain(board: Board, square: Square, rank: number, file: number, color: Color): boolean {
  const behindRank = color === "white" ? Math.max(0, rank - 1) : Math.min(7, rank + 1);
  
  const adjacentFiles = [Math.max(0, file - 1), Math.min(7, file + 1)];
  for (const adjacentFile of adjacentFiles) {
    if (adjacentFile !== file) {
      const checkSquare = behindRank * 8 + adjacentFile;
      const piece = board.getPiece(checkSquare);
      if (piece && piece.color === color && piece.type === "P") {
        return true;
      }
    }
  }
  
  return false;
}

function isBackward(board: Board, square: Square, rank: number, file: number, color: Color, pawnFiles: number[]): boolean {
  const leftFile = Math.max(0, file - 1);
  const rightFile = Math.min(7, file + 1);
  
  const adjacentFiles = [leftFile, rightFile];
  for (const adjacentFile of adjacentFiles) {
    if (adjacentFile !== file && pawnFiles[adjacentFile] > 0) {
      for (let checkSquare = 0; checkSquare < 64; checkSquare++) {
        const piece = board.getPiece(checkSquare);
        if (piece && piece.color === color && piece.type === "P") {
          const checkFile = checkSquare % 8;
          const checkRank = Math.floor(checkSquare / 8);
          
          if (checkFile === adjacentFile) {
            const isAhead = color === "white" ? checkRank > rank : checkRank < rank;
            
            if (isAhead) {
              return false;
            }
          }
        }
      }
    }
  }
  
  return false;
}

File: implementations/rust/src/eval/pawn_structure.rs
use crate::board::Board;
use crate::types::{Color, PieceType, Square};

const PASSED_PAWN_BONUS: [i32; 8] = [0, 10, 20, 40, 60, 90, 120, 0];
const DOUBLED_PAWN_PENALTY: i32 = -20;
const ISOLATED_PAWN_PENALTY: i32 = -15;
const BACKWARD_PAWN_PENALTY: i32 = -10;
const CONNECTED_PAWN_BONUS: i32 = 5;
const PAWN_CHAIN_BONUS: i32 = 10;

pub fn evaluate(board: &Board) -> i32 {
    let mut score = 0;
    
    score += evaluate_color(board, Color::White);
    score -= evaluate_color(board, Color::Black);
    
    score
}

fn evaluate_color(board: &Board, color: Color) -> i32 {
    let mut score = 0;
    let mut pawn_files = [0u8; 8];
    let mut pawn_positions = Vec::new();
    
    for square in 0..64 {
        if let Some(piece) = board.get_piece(square) {
            if piece.color == color && piece.piece_type == PieceType::Pawn {
                let file = square % 8;
                let rank = square / 8;
                pawn_files[file as usize] += 1;
                pawn_positions.push((square, rank, file));
            }
        }
    }
    
    for (square, rank, file) in pawn_positions.iter() {
        if pawn_files[*file as usize] > 1 {
            score += DOUBLED_PAWN_PENALTY;
        }
        
        if is_isolated(*file, &pawn_files) {
            score += ISOLATED_PAWN_PENALTY;
        }
        
        if is_passed(board, *square, *rank, *file, color) {
            let bonus_rank = if color == Color::White { *rank } else { 7 - *rank };
            score += PASSED_PAWN_BONUS[bonus_rank as usize];
        }
        
        if is_connected(board, *square, *file, color) {
            score += CONNECTED_PAWN_BONUS;
        }
        
        if is_in_chain(board, *square, *rank, *file, color) {
            score += PAWN_CHAIN_BONUS;
        }
        
        if is_backward(board, *square, *rank, *file, color, &pawn_files) {
            score += BACKWARD_PAWN_PENALTY;
        }
    }
    
    score
}

fn is_isolated(file: Square, pawn_files: &[u8; 8]) -> bool {
    let left_file = if file > 0 { pawn_files[(file - 1) as usize] } else { 0 };
    let right_file = if file < 7 { pawn_files[(file + 1) as usize] } else { 0 };
    left_file == 0 && right_file == 0
}

fn is_passed(board: &Board, square: Square, rank: Square, file: Square, color: Color) -> bool {
    let (start_rank, end_rank, direction) = if color == Color::White {
        (rank + 1, 8, 1)
    } else {
        (0, rank, -1)
    };
    
    for check_file in (file.saturating_sub(1))..=(file + 1).min(7) {
        let mut current_rank = start_rank;
        loop {
            if color == Color::White {
                if current_rank >= end_rank { break; }
            } else {
                if current_rank >= rank { break; }
            }
            
            let check_square = current_rank * 8 + check_file;
            if let Some(piece) = board.get_piece(check_square) {
                if piece.piece_type == PieceType::Pawn && piece.color != color {
                    return false;
                }
            }
            
            current_rank = if direction > 0 { current_rank + 1 } else { current_rank.saturating_sub(1) };
            if direction < 0 && current_rank == 0 { break; }
        }
    }
    
    true
}

fn is_connected(board: &Board, square: Square, file: Square, color: Color) -> bool {
    let rank = square / 8;
    
    for adjacent_file in [file.saturating_sub(1), (file + 1).min(7)].iter() {
        if *adjacent_file != file {
            let adjacent_square = rank * 8 + adjacent_file;
            if let Some(piece) = board.get_piece(adjacent_square) {
                if piece.color == color && piece.piece_type == PieceType::Pawn {
                    return true;
                }
            }
        }
    }
    
    false
}

fn is_in_chain(board: &Board, square: Square, rank: Square, file: Square, color: Color) -> bool {
    let behind_rank = if color == Color::White {
        rank.saturating_sub(1)
    } else {
        (rank + 1).min(7)
    };
    
    for adjacent_file in [file.saturating_sub(1), (file + 1).min(7)].iter() {
        if *adjacent_file != file {
            let check_square = behind_rank * 8 + adjacent_file;
            if let Some(piece) = board.get_piece(check_square) {
                if piece.color == color && piece.piece_type == PieceType::Pawn {
                    return true;
                }
            }
        }
    }
    
    false
}

fn is_backward(board: &Board, square: Square, rank: Square, file: Square, color: Color, pawn_files: &[u8; 8]) -> bool {
    let left_file = file.saturating_sub(1);
    let right_file = (file + 1).min(7);
    
    for adjacent_file in [left_file, right_file].iter() {
        if *adjacent_file != file && pawn_files[*adjacent_file as usize] > 0 {
            for check_square in 0..64 {
                if let Some(piece) = board.get_piece(check_square) {
                    if piece.color == color && piece.piece_type == PieceType::Pawn {
                        let check_file = check_square % 8;
                        let check_rank = check_square / 8;
                        
                        if check_file == *adjacent_file {
                            let is_ahead = if color == Color::White {
                                check_rank > rank
                            } else {
                                check_rank < rank
                            };
                            
                            if is_ahead {
                                return false;
                            }
                        }
                    }
                }
            }
        }
    }
    
    false
}
